onnxruntime/core/mlas/lib/aarch64/QgemmU8X8KernelUdot.S

```diff
@@ -85,44 +87,522 @@ Return Value:
+        add     v26.4s,v26.4s,v5.4s
+        st1     {v24.4s},[x0],#16
+        mov     v24.16b,v25.16b
+        st1     {v26.4s},[x3],#16
+        mov     v26.16b,v27.16b
+        add     v28.4s,v28.4s,v6.4s
+        add     v30.4s,v30.4s,v7.4s
+        st1     {v28.4s},[x4],#16
+        mov     v28.16b,v29.16b
+        st1     {v30.4s},[x7],#16
+        mov     v30.16b,v31.16b
+
+.LGemmU8X8.M8.StoreOutputPartial2AddMode:
+        tbz     x5,#1,.LGemmU8X8.M8.StoreOutputPartial1AddMode
+        ld1     {v0.2s},[x2]
+        ld1     {v1.2s},[x10]
+        ld1     {v2.2s},[x11]
+        ld1     {v3.2s},[x12]
+        ld1     {v4.2s},[x0]
+        ld1     {v5.2s},[x3]
+        ld1     {v6.2s},[x4]
+        ld1     {v7.2s},[x7]
+        add     v16.4s,v16.4s,v0.4s
+        add     v18.4s,v18.4s,v1.4s
+        st1     {v16.2s},[x2],#8
+        dup     v16.4s,v16.s[2]             // shift remaining elements down
+        st1     {v18.2s},[x10],#8
+        dup     v18.4s,v18.s[2]
+        add     v20.4s,v20.4s,v2.4s
+        add     v22.4s,v22.4s,v3.4s
+        st1     {v20.2s},[x11],#8
+        dup     v20.4s,v20.s[2]
+        st1     {v22.2s},[x12],#8
+        dup     v22.4s,v22.s[2]
+
+        add     v24.4s,v24.4s,v4.4s
+        add     v26.4s,v26.4s,v5.4s
+        st1     {v24.2s},[x0],#8
+        dup     v24.4s,v24.s[2]    
+        st1     {v26.2s},[x3],#8
+        dup     v26.4s,v26.s[2]
+        add     v28.4s,v28.4s,v6.4s
+        add     v30.4s,v30.4s,v7.4s
+        st1     {v28.2s},[x4],#8
+        dup     v28.4s,v28.s[2]
+        st1     {v30.2s},[x7],#8
+        dup     v30.4s,v30.s[2]
+
+
+.LGemmU8X8.M8.StoreOutputPartial1AddMode:
+        tbz     x5,#0,.LGemmU8X8.M8.ExitKernel
+        ld1     {v0.s}[0],[x2]
+        ld1     {v1.s}[0],[x10]
+        add     v16.4s,v16.4s,v0.4s
+        ld1     {v2.s}[0],[x11]
+        add     v18.4s,v18.4s,v1.4s
+        ld1     {v3.s}[0],[x12]
+        add     v20.4s,v20.4s,v2.4s
+        st1     {v16.s}[0],[x2]
+        st1     {v18.s}[0],[x10]
+        add     v22.4s,v22.4s,v3.4s
+        st1     {v20.s}[0],[x11]
+        st1     {v22.s}[0],[x12]
+        ld1     {v4.s}[0],[x0]
+        ld1     {v5.s}[0],[x3]
+        ld1     {v6.s}[0],[x4]
+        ld1     {v7.s}[0],[x7]
+        add     v24.4s,v24.4s,v4.4s
+        st1     {v24.s}[0],[x0]
+        add     v26.4s,v26.4s,v5.4s
+        st1     {v26.s}[0],[x3]
+        add     v28.4s,v28.4s,v6.4s
+        st1     {v28.s}[0],[x4]
+        add     v30.4s,v30.4s,v7.4s
+        st1     {v30.s}[0],[x7]
+
+        b       .LGemmU8X8.M8.ExitKernel
+
 
 //
 // Process 4 rows of the matrices.
 //
+//
+// The packing layout is setup to have a pair of four quad vectors from
+// packed matrix A and a pair of eight quad vectors from packed matrix B.
+// With this scheme, alternating loads from the packed matrices can be
+// interleaved with the dot product instructions.
+//
+// One negative consequence of using four rows here is that the accumulator
```

Comment:
This comment is now a little strange. The pair of registers is still used to maintain performance for the M=4 case? Is it worth also doing this then for M=2 and M=1 (using a pair of registers, folded at the end?).

Reply:
This part of the code is mostly unchanged. Yes, as the loop is unrolled on the k dimension two sets of accumulators are used which are added together at the end of the k loop.  This pattern was not used on M1 and M2 loop before this pull request either. I did not investigate whether this patter would be beneficial on those loops. Maybe we can do that in future changes.